### ISSUE TITLE
There's no license

### ISSUE THREAD

**Issue by turol** [khedoros/uw-engine#8]
Your code doesn't have any kind of license. In absence of one it's not possible for anyone else to touch this in any way. Blame stupid copyright laws.

Pick an open-source license, add it to the project and the appropriate header to all source files. I prefer GPL3+ for non-library projects which are not meant to used in other projects and MIT for libraries. I think GitHub has a page for common licenses. If not, head over to [Open Source Initiative](https://opensource.org/licenses).

**Comment by khedoros** [comment]
I added a copy of the GPL3 license to the root of the project, but haven't put copyright headers on the source; I'm going to need to work out some issues for at least some of the files.

Most of the files are of my own authorship, and those aren't any issue. Some files are of uncertain provenance, and I'll have to figure out where I got them from, the exact license that would apply, and so on. The biggest issues will be the OPL (Yamaha FM synthesis chips) emulators. I suspect that they're from some version of MAME/MESS, or possibly from one of the Doom source ports (which themselves took from MAME/MESS...I keep coming back to the arcade emulators, because I'm pretty sure they're involved somehow, even if indirectly).

Anyhow, audio's the biggest mess, but the interface to switch to a new OPL emulator is pretty easy, and I'm inclined to just delete implementations that I have unclear rights to use. Worst-case: one of these is a C++ port of an LGPL-licensed Java OPL3 emulator, and I could go and port that myself. The stuff from John Miles (also noted below) has a less tried-and-true license, but it reads to me as "I want this code to be public domain"; I included the text of the readme.txt file at the bottom of this comment.

These are the files that I didn't write, and the licensing statuses for each, as far as I can remember them (keeping in mind that a number of things were last touched in 2015 and 2016):

**FreeSans.ttf** is from the GNU FreeFont project, and is licensed under GPLv3+, but with an [added exception](https://www.gnu.org/software/freefont/license.html) involving embedding the font in a document.

**audio/opl/OPL3.cpp** is Copyright Robson Cozendey, under an LGPL 2.1+ license, but that refers to the Java version. I'm not sure who did the port to C++, but I believe that the extra work would have to fall under the original license, as a derivative work.

**audio/opl/opl.h** is from the same source as that emulator, if I'm remembering correctly. It defines an abstract interface for the use of different OPL emulators in the project, and it's the interface that both OPL3.cpp and fmopl.cpp have been modified to use.

**audio/opl/fmopl.cpp** is "based on fmopl.c from MAME 0.95", and I'm not sure if I made modifications to wrap it in a C++ class, or if that was someone else (and if it was someone else, then whom). OPL3.cpp has a cleaner license, and this is just an alternate implementation of an OPL emulator, so it might make sense to just remove this from the project for the sake of clarity. The file contains some information relevant to licensing at the top of the file. It requires that it be put to no commercial use, but encourages use in non-commercial projects, and distribution of the source. These are murky-seeming terms.

**audio/opl/xs_float.h** is...also probably from the source of the past 3 files, which it's becoming clear I'll have to track down. It's originally copyright Sree Kotay, 2006, and doesn't have a clear license attached to it.

All of the uppercase filenames in **audio/kail/[A-Z]**: These are from John Miles's AIL2 files, open-sourced in 2000. These are the license terms provided in his readme:

```
This package contains the real mode (AIL V2.14) and 32-bit protected-mode DOS
(AIL/32 V1.05) versions of the IBM Audio Interface Library (AIL).  AIL V2.14
and AIL/32 were the predecessors to the current Miles Sound System package
published by RAD Game Tools (http://www.radgametools.com).

As of May 26, 2000, all source code and compiled object code in this package
may be considered "open-source freeware," usable by anyone for any purpose,
commercial or otherwise, without restriction or limitation.  I strongly
encourage, but do not require, anyone who adds any further value to this
package to release their modifications and improvements under the same terms.

I expressly disclaim any liability for bugs, defects, or intellectual-property
violations that may arise as a consequence of using the contents of this
package for any purpose, commercial or otherwise.  Use at your own risk!

John Miles (jmiles@pop.net)
Miles Design, Inc.
May 26, 2000
```

**Comment by turol** [comment]
This is not much better.

1. GPL forbids linking with proprietary code. If you have any code of uncertain origin it's still impossible to touch this.

2. If a file doesn't have a clear license, you're not allowed to distribute/use it and this whole repository is forbidden.

3. I'm not a lawyer so I can't tell if the Miles sound system license is open source or GPL-compatible. You need to either consult with a lawyer or remove it.

4. If you use stuff written by @hankmorgan or that he might want to use in his project you need to discuss licensing with him.

Things like this are why open source releases by big companies can take months or years. Copyright law *sucks*.

**Comment by khedoros** [comment]
Nope, not much better, and thus the issue is still open. Just a clearer statement of where things stand. 

The project doesn't have a budget; as a hobby project that I'm not willing to legally defend, if any legal trouble pops up, it's going to disappear. 

I've written to John Miles in the past, talking about plans for this project. I believe I've exchanged messages with Hank as well, and determined that our strategies were essentially incompatible. His project looks to be far ahead, anyhow. My project doesn't include any of his code. I mentioned his project merely as a point of interest; someone interested in this project would likely also be interested in another, related project.

John Miles' code is in the repo as reference. It can be easily removed from the repo. They may not be verifiable as GPL-compatible without legal work that just isn't going to happen.

The OPL code may be as simple as recursive documentation of the provenance of the code, and the license that it was distributed under at each level. Or it might mean ripping it out and finding a clearer-licensed replacement.

**Comment by khedoros** [comment]
Found the licenses for the 2 OPL emulators that I use and the OPL interface that I derived mine from, and documented them in my LICENSE file. Removed the AIL code, since it was unclear to me how the license would taint the rest of the code in the repo. The rest of the files in https://github.com/khedoros/uw-engine/tree/master/audio/kail are authored by me.

audio/opl/xs_float.h is a possible sticking point still. It's used in one of the OPL emulators. ZDoom claims that they can use it under a GPL-compatible license, but [the original author](http://stereopsis.com/sree/fpu2006.html) doesn't specify a license for the code that they posted, and I can't see anywhere obvious where they asked the author to use the code, so I think they're wrong. Luckily, it's just used for some speed-optimized floor/ceil/round functions, and only in a couple of places. Actually removing it will be a project for tonight perhaps.
